### PR TITLE
Further synchronization

### PR DESCRIPTION
**Pull request by ovalseven8**
After skimming through the code I believe further synchronization is necessary to avoid lost updates or other concurrency issues.

### CODE REVIEW

**Comment by dbrgn** [comment]
Thanks! I'll take a look tomorrow if I find time.

Keep https://github.com/saltyrtc/saltyrtc-client-java/issues/114 in mind! So only things need to be synchronized that are being accessed concurrently *within* the library itself.

File: src/main/java/org/saltyrtc/client/signaling/Signaling.java [hunk 1]
```diff
@@ -294,20 +294,18 @@ private void initWebsocket() throws IOException {
...
             @SuppressWarnings("UnqualifiedMethodAccess")
-            public void onConnected(WebSocket websocket, Map<String, List<String>> headers) {
-                synchronized (this) {
-                    if (getState() == SignalingState.WS_CONNECTING) {
-                        getLogger().info("WebSocket connection established");
-                        setState(SignalingState.SERVER_HANDSHAKE);
-                    } else {
-                        getLogger().warn("Got onConnected event, but WebSocket connection already open");
-                    }
+            public synchronized void onConnected(WebSocket websocket, Map<String, List<String>> headers) {
```
Comment (ovalseven8) [inline src/main/java/org/saltyrtc/client/signaling/Signaling.java]: Just a syntax change here.

**Comment by ovalseven8** [comment]
Some reasoning:
The callbacks are being executed by different threads in the current implementation (at least `ReadingThread` and `WritingThread`). The callbacks can run in parallel as far as I understand.

**Why did I add `synchronized` inside `onDisconnected()` and other callbacks?**
AFAIK it's possible that `onDisconnected()` is executed by at least both `ReadingThread` and `WritingThread`. So, without this change it's possible that `onDisconnected()` and `onBinaryMessage()` run in parallel. Both access shared state. Lost updates and race conditions should be possible.
That's also the reason why `onConnected()`, `onConnectError()` and `onTextMessage()` do need the synchronization IMHO.

**Why did I add `synchronized` for `onCallbackError()`?**
The method calls internally the synchronized method `resetConnection()`, however, `resetConnection()` is synchronized using the `Signaling` class instance while we also need synchronization on the `WebSocketAdapter` class instance. Because we have to guarantee a "happens before" relationship, only this way the JVM makes sure that the thread uses the latest changes on variables like `this.state`. Otherwise, "lost updates" are possible. I am not sure if we need synchronization for `resetConnection()` anymore then.

File: src/main/java/org/saltyrtc/client/signaling/Signaling.java [hunk 2]
```diff
@@ -502,9 +500,11 @@ public void onDisconnected(WebSocket websocket,
                     }
                 }
                 // Note: Don't check for signaling state here, it will already have been resetted.
-                if (closeCode != CloseCode.HANDOVER) {
-                    Signaling.this.salty.events.close.notifyHandlers(new CloseEvent(closeCode));
-                    setState(SignalingState.CLOSED);
+                synchronized (this) {
```
Comment (lgrahl) [inline src/main/java/org/saltyrtc/client/signaling/Signaling.java]: Let's just make the whole method synchronised.

**Comment by ovalseven8** [comment]
> That being said, we still face many potential race conditions since the application uses its own thread, so this only fixes potential races between the WebSocket threads.

You mean in combination with e.g. `saltyrtc-task-webrtc-java`? Or even in `saltyrtc-client-java` alone?

File: src/main/java/org/saltyrtc/client/signaling/Signaling.java (hunk 2 above)
Comment (ovalseven8) [reply src/main/java/org/saltyrtc/client/signaling/Signaling.java]: Done.

**Comment by lgrahl** [comment]
The latter.

**Pull request merged**